File: bridges/nodejs/src/sdk/tools/ecapa-tool.ts
import fs from 'node:fs'

import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'

const MODEL_NAME = 'ecapa-voice_gender_classifier'

export default class ECAPATool extends Tool {
  private static readonly TOOLKIT = 'music_audio'
  private readonly config: ReturnType<typeof ToolkitConfig.load>

  constructor() {
    super()
    // Load configuration from central toolkits directory
    this.config = ToolkitConfig.load(ECAPATool.TOOLKIT, this.toolName)
  }

  get toolName(): string {
    // Use the actual config name for toolkit lookup
    return 'ecapa'
  }

  get toolkit(): string {
    return ECAPATool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Detect gender from audio file using ECAPA-TDNN voice gender classifier
   * @param inputPath The file path of the audio to be analyzed
   * @param device Device to use for processing (cpu, cuda)
   * @returns A promise that resolves with the detected gender: "male", "female", or "unknown"
   */
  async detectGender(inputPath: string, device = 'cpu'): Promise<string> {
    try {
      // Validate input file exists
      if (!fs.existsSync(inputPath)) {
        throw new Error(`Input file does not exist: ${inputPath}`)
      }

      // Get model path using the generic resource system
      const modelPath = await this.getResourcePath(MODEL_NAME)

      const args = [
        '--function',
        'detect_gender',
        '--input',
        inputPath,
        '--model_path',
        modelPath,
        '--device',
        device
      ]

      const result = await this.executeCommand({
        binaryName: 'ecapa-voice_gender_classifier',
        args,
        options: { sync: true }
      })

      // Parse the output to extract gender
      const gender = this.parseGenderOutput(result)

      return gender
    } catch (error: unknown) {
      throw new Error(
        `Voice gender detection failed: ${(error as Error).message}`
      )
    }
  }

  /**
   * Parse the gender detection output
   */
  private parseGenderOutput(rawOutput: string): string {
    const lines = rawOutput.split('\n')

    // Look for gender result in the output
    for (const line of lines) {
      const lowerLine = line.toLowerCase().trim()

      if (lowerLine.includes('gender:')) {
        // Extract gender from line like "Gender: male"
        const match = lowerLine.match(/gender:\s*(male|female|unknown)/i)
        if (match && match[1]) {
          return match[1].toLowerCase()
        }
      }

      // Also check for direct gender output
      if (lowerLine === 'male' || lowerLine === 'female') {
        return lowerLine
      }
    }

    // If no clear gender found, return unknown
    return 'unknown'
  }
}

File: bridges/python/src/sdk/tools/ecapa_tool.py
import os

from ..base_tool import BaseTool, ExecuteCommandOptions
from ..toolkit_config import ToolkitConfig

MODEL_NAME = 'ecapa-voice_gender_classifier'


class ECAPATool(BaseTool):
    """
    Tool for voice gender classification using ECAPA-TDNN model.
    
    Example output format:
    Gender: male
    """

    TOOLKIT = 'music_audio'

    def __init__(self):
        super().__init__()
        # Load configuration from central toolkits directory
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)

    @property
    def tool_name(self) -> str:
        # Use the actual config name for toolkit lookup
        return 'ecapa'

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config['description']

    def detect_gender(
        self,
        input_path: str,
        device: str = 'cpu'
    ) -> str:
        """
        Detect gender from audio file using ECAPA-TDNN voice gender classifier

        Args:
            input_path: The file path of the audio to be analyzed
            device: Device to use for processing (cpu, cuda)

        Returns:
            The detected gender: "male", "female", or "unknown"
        """
        try:
            # Validate input file exists
            if not os.path.exists(input_path):
                raise Exception(f"Input file does not exist: {input_path}")

            # Get model path using the generic resource system
            model_path = self.get_resource_path(MODEL_NAME)

            args = [
                '--function', 'detect_gender',
                '--input', input_path,
                '--model_path', model_path,
                '--device', device
            ]

            result = self.execute_command(ExecuteCommandOptions(
                binary_name='ecapa-voice_gender_classifier',
                args=args,
                options={'sync': True}
            ))

            # Parse the output to extract gender
            gender = self._parse_gender_output(result)

            return gender
        except Exception as e:
            raise Exception(f"Voice gender detection failed: {str(e)}")

    def _parse_gender_output(self, raw_output: str) -> str:
        """
        Parse the gender detection output
        
        Args:
            raw_output: Raw output from the gender detection binary
            
        Returns:
            Detected gender: "male", "female", or "unknown"
        """
        lines = raw_output.split('\n')

        # Look for gender result in the output
        for line in lines:
            lower_line = line.lower().strip()

            if 'gender:' in lower_line:
                # Extract gender from line like "Gender: male"
                import re
                match = re.search(r'gender:\s*(male|female|unknown)', lower_line, re.IGNORECASE)
                if match:
                    return match.group(1).lower()

            # Also check for direct gender output
            if lower_line in ['male', 'female']:
                return lower_line

        # If no clear gender found, return unknown
        return 'unknown'
